### PR TITLE
[MLIR] Remove redundant solver local applicable checks

### PR DESCRIPTION
**Pull request by jerryyin**
I had a discussion with @zjing14 and @asroy and get the suggestion we set the single point of truth in judging applicability in MLIR instead of the existing hybrid (MLIR + MIOpen approach). The reasoning is that the applicability check in MIOpen end may grow out-of-sync with the actual MLIR solver capabilities. (One example of this is that EPack feature has not been implemented in MLIR but yet we have this redundant check `GetEPackLength()` that undermine the MLIR config coverage.)

In the PR, I'm removing all non-trivial applicability checks for MLIR backend solvers. Instead, the check will be performed by the pre-existing `MiirIsConfigApplicable()` check: This check will:
 - Invoke a MLIR API that does the applicability check
 - Lower through the first 3 step of passes: MIOpenOps, AffineTransform and AffixTuningParams
 - Config is applicable if and only if the lowering succeed: meaning the algorithm is able to find a compatible set of tuning parameters and proceed.

For this PR, I want all reviewers to approve before merging.

### CODE REVIEW

**Comment by atamazov** [comment]
> Instead, the check will be performed by the pre-existing MiirIsConfigApplicable() check: This check will...

Is this done already?

**Comment by jerryyin** [comment]
> Is this done already?

Yes, this is already done. Therefore removing the redundant applicable checks are safe at this point.

**Review by atamazov — DISMISSED** [review]
LGTM!

**Comment by atamazov** [comment]
`on_hold` because
> For this PR, I want ___all reviewers to approve___ before merging.

File: src/solver/conv_mlir_igemm_fwd.cpp
Comment (asroy) [inline src/solver/conv_mlir_igemm_fwd.cpp:66]: This should also be moved into MLIR

File: src/solver/conv_mlir_igemm_wrw.cpp
Comment (asroy) [inline src/solver/conv_mlir_igemm_wrw.cpp:69]: Same as above, and below (if there is any)

**Review by asroy — DISMISSED** [review]

**Review by whchung — CHANGES_REQUESTED** [review]
Per discussion, MLIR kernel generator still needs validations, especially when it comes to cases where Gemm M/N/K dimensions have to be extended.

for example:

- C=3 config in resnet50
- some configs in resnext101 , especially in corner cases where the extent we enlarge along Gemm M / N dimensions are different.

Before that I don't think it's a good idea to merge this PR.

**Comment by jerryyin** [comment]
Converted to draft. Feedbacks from @whchung and @asroy make me realize that we are probably missing this feature in MLIR now. Until we can correctly reject non-supported configs from MLIR, the checks have to be in place in MIOpen.

File: src/solver/conv_mlir_igemm_fwd.cpp
Comment (jerryyin) [reply src/solver/conv_mlir_igemm_fwd.cpp:66]: Conceptually all applicability check can be moved to let mlir handle it. However, that means every time MIOpen check for applicability of any config (like bf16), will ask MLIR's opinion. And the performance overhead of that check would be lowering through the first 3 steps of the pipeline.

My suggestion is that we try to hit at a balance between effectiveness and performance. Right now let's leave those extremely cheap checks in MIOpen untouched. Once we make the query overhead in checking with MLIR much lower than now, we can consider to remove those.

**Comment by jerryyin** [comment]
Quick update to all: 

For the short term we will do the following:
  @asleepzzz  will take over the branch and add necessary checks for padding kernel.

For the mid/long term, we will shoot for removing applicable checks entirely.

**Review by whchung — APPROVED** [review]

**Comment by asleepzzz** [comment]
it seems that fwd and wrw walk through different path, 
even remove limitations of mlir solver

bin/MIOpenDriver convfp16 -F 4 -n 256 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
blocked by other condition

**Comment by asleepzzz** [comment]
resnet50 configs passed


./bin/MIOpenDriver convfp16 -F 4 -n 256 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
MIOpenDriver convfp16 -F 4 -n 256 -c 3 -H 224 -W 224 -k 64 -y 7 -x 7 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
MIOpen Backward Weights Conv. Algorithm: 5, Solution: 105/ConvMlirIgemmWrWXdlops
GPU Kernel Time Backward Weights Conv. Elapsed: 2199.225830 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: bwdw-conv7x7u2, 256, 3, 112, 112, 7, 7, 64,  60423143424, 0, 0, 27, 0, 2199.225830
Backward Convolution Weights Verifies OK on CPU reference (0.000225131)


./bin/MIOpenDriver convfp16 -F 1 -n 256 -c 3 -H 230 -W 230 -k 64 -y 7 -x 7 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
MIOpenDriver convfp16 -F 1 -n 256 -c 3 -H 230 -W 230 -k 64 -y 7 -x 7 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
MIOpen Forward Conv. Algorithm: 5, Solution: 103/ConvMlirIgemmFwdXdlops
GPU Kernel Time Forward Conv. Elapsed: 4066.726074 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv7x7u2, 256, 3, 112, 112, 7, 7, 64,  60423143424, 81273216, 411041792, 15, 0, 4066.726074
Forward Convolution Verifies OK on CPU reference (9.17173e-05)

**Pull request merged**
